File: LeetCode/249.Group_Shifted_Strings/249.Group_Shifted_Strings.py
#!/usr/bin/env python3


class Solution:
    def groupStrings(self, strings: List[str]) -> List[List[str]]:

        def get_key(s):
            res = ""
            for i in range(1, len(s)):
                dis = (ord(s[i]) - ord(s[i - 1])) % 26
                res += chr(dis + ord("a"))
            return res

        dic = defaultdict(list)

        for s in strings:
            key = get_key(s)
            dic[key].append(s)

        # print(dic)

        return [dic[k] for k in dic]

File: LeetCode/249.Group_Shifted_Strings/249.Group_Shifted_Strings.swift
class Solution {
    func groupStrings(_ strings: [String]) -> [[String]] {
        // right shift: a->b
        // left shift: a -> z
        // abc -> 123

        var buckets: [String: [String]] = [:]
        for s in strings {
            let key = getKey(s)
            // print(key)
            buckets[key, default: []].append(s)
        }

        var res: [[String]] = []
        for v in buckets.values {
            res.append(v)
        }

        return res
    }

    func getKey(_ s: String) -> String {
        let bytes = Array(s.utf8).map { Int($0) }
        let n = bytes.count
        guard n > 1 else { return "#" }

        var parts: [String] = []
        for i in 1 ..< n {
            let diff = ((bytes[i] - bytes[i - 1]) + 26) % 26
            parts.append(String(diff))
        }

        return parts.joined(separator: "#")
    }
}
